File: packages/core/src/imports.ts
// Import graph detection via regex — no native dependencies

import type { CodeFile } from "./types.js"

const IMPORT_PATTERNS: Record<string, RegExp> = {
  typescript: /(?:import|from)\s+['"](\.[^'"]+)['"]/g,
  javascript: /(?:import|from)\s+['"](\.[^'"]+)['"]/g,
  go: /import\s+(?:[\w]+\s+)?["'](\.[^"']+)["']/g,
  python: /from\s+(\.[^\s]+)\s+import/g,
  rust: /(?:use|mod)\s+(super|self)::[^\s;]+/g,
}

export type ImportGraph = Map<string, string[]>

/**
 * Build a relative import graph for a list of files.
 */
export function buildImportGraph(files: CodeFile[]): ImportGraph {
  const graph: ImportGraph = new Map()

  for (const file of files) {
    const pattern = IMPORT_PATTERNS[file.language]
    if (!pattern) continue

    const imports: string[] = []
    pattern.lastIndex = 0

    let match: RegExpExecArray | null
    while ((match = pattern.exec(file.content)) !== null) {
      const imp = match[1]
      if (imp) imports.push(imp)
    }

    if (imports.length > 0) {
      graph.set(file.path, imports)
    }
  }

  return graph
}

const BOOST_FACTOR = 0.3

/**
 * Calculate import boost scores for files imported by top-scored files.
 */
export function calculateImportBoosts(
  graph: ImportGraph,
  scores: Map<string, number>,
): Map<string, number> {
  const boosts = new Map<string, number>()

  for (const [importer, imported] of graph) {
    const importerScore = scores.get(importer) ?? 0
    if (importerScore <= 0) continue

    for (const dep of imported) {
      const current = boosts.get(dep) ?? 0
      boosts.set(dep, current + importerScore * BOOST_FACTOR)
    }
  }

  return boosts
}

File: go/internal/trimmer/imports.go
package trimmer

import (
	"context"
	"path/filepath"
	"strings"

	sitter "github.com/smacker/go-tree-sitter"
	"github.com/smacker/go-tree-sitter/golang"
	"github.com/smacker/go-tree-sitter/javascript"
	"github.com/smacker/go-tree-sitter/python"
	"github.com/smacker/go-tree-sitter/typescript/typescript"
)

// ImportGraph maps file paths to the files they import.
type ImportGraph map[string][]string

// BuildImportGraph analyzes all files and returns their import relationships.
func BuildImportGraph(files []CodeFile) ImportGraph {
	graph := make(ImportGraph)
	for _, f := range files {
		imports := extractImports(f)
		if len(imports) > 0 {
			graph[f.Path] = imports
		}
	}
	return graph
}

func extractImports(file CodeFile) []string {
	lang := detectLanguage(file)
	if lang == nil {
		return nil
	}

	parser := sitter.NewParser()
	parser.SetLanguage(lang)

	tree, err := parser.ParseCtx(context.Background(), nil, []byte(file.Content))
	if err != nil || tree == nil {
		return nil
	}

	switch strings.ToLower(file.Language) {
	case "go", "":
		if filepath.Ext(file.Path) == ".go" {
			return extractGoImports(tree, file.Content)
		}
	case "python", "py":
		return extractPythonImports(tree, file.Content)
	case "typescript", "ts":
		return extractJSImports(tree, file.Content)
	case "javascript", "js":
		return extractJSImports(tree, file.Content)
	}

	switch filepath.Ext(file.Path) {
	case ".go":
		return extractGoImports(tree, file.Content)
	case ".py":
		return extractPythonImports(tree, file.Content)
	case ".ts":
		return extractJSImports(tree, file.Content)
	case ".js":
		return extractJSImports(tree, file.Content)
	}
	return nil
}

func detectLanguage(file CodeFile) *sitter.Language {
	switch strings.ToLower(file.Language) {
	case "go":
		return golang.GetLanguage()
	case "python", "py":
		return python.GetLanguage()
	case "typescript", "ts":
		return typescript.GetLanguage()
	case "javascript", "js":
		return javascript.GetLanguage()
	}

	switch filepath.Ext(file.Path) {
	case ".go":
		return golang.GetLanguage()
	case ".py":
		return python.GetLanguage()
	case ".ts":
		return typescript.GetLanguage()
	case ".js":
		return javascript.GetLanguage()
	}
	return nil
}

func extractGoImports(tree *sitter.Tree, content string) []string {
	var imports []string
	src := []byte(content)

	var walk func(node *sitter.Node)
	walk = func(node *sitter.Node) {
		if node.Type() == "interpreted_string_literal" {
			parent := node.Parent()
			if parent != nil && parent.Type() == "import_spec" {
				path := strings.Trim(string(node.Content(src)), `"`)
				if strings.HasPrefix(path, "./") || strings.HasPrefix(path, "../") {
					imports = append(imports, path)
				}
			}
		}
		for i := 0; i < int(node.ChildCount()); i++ {
			walk(node.Child(i))
		}
	}
	walk(tree.RootNode())
	return imports
}

func extractPythonImports(tree *sitter.Tree, content string) []string {
	var imports []string
	src := []byte(content)

	var walk func(node *sitter.Node)
	walk = func(node *sitter.Node) {
		if node.Type() == "relative_import" {
			imports = append(imports, string(node.Content(src)))
		}
		for i := 0; i < int(node.ChildCount()); i++ {
			walk(node.Child(i))
		}
	}
	walk(tree.RootNode())
	return imports
}

func extractJSImports(tree *sitter.Tree, content string) []string {
	var imports []string
	src := []byte(content)

	var walk func(node *sitter.Node)
	walk = func(node *sitter.Node) {
		if node.Type() == "import_statement" {
			for i := 0; i < int(node.ChildCount()); i++ {
				child := node.Child(i)
				if child.Type() == "string" {
					path := strings.Trim(string(child.Content(src)), `"'`)
					if strings.HasPrefix(path, ".") {
						imports = append(imports, path)
					}
				}
			}
		}
		for i := 0; i < int(node.ChildCount()); i++ {
			walk(node.Child(i))
		}
	}
	walk(tree.RootNode())
	return imports
}

// ImportBoost returns a boost score for files imported by top-scored files.
func ImportBoost(graph ImportGraph, scores map[string]float64) map[string]float64 {
	boosts := make(map[string]float64)
	const boostFactor = 0.3

	for importer, imported := range graph {
		importerScore := scores[importer]
		if importerScore <= 0 {
			continue
		}
		for _, dep := range imported {
			boosts[dep] += importerScore * boostFactor
		}
	}
	return boosts
}
